Core/Source/MVVM/NestedModelCollection.swift

```diff
@@ -0,0 +1,50 @@
+import Foundation
+
+/// Specialization of a `ModelCollection` that provides support for nesting - that is having multiple levels of models
+/// for example when displaying in an outline or browser view.
+public protocol NestedModelCollection: ModelCollection {
+    func isModelExpandable(_ model: Model) -> Bool
+    func childModelCollection(_ model: Model) -> NestedModelCollection
+}
+
+public extension ModelCollection {
+    public func asNested() -> NestedModelCollection {
+        if let nested = self as? NestedModelCollection { return nested }
+        return SingleLevelNestedModelCollection(self)
+    }
+}
+
+/// Internal only class used for wrapping non-nested `ModelCollection` into a nested model collection.
+internal final class SingleLevelNestedModelCollection: NestedModelCollection {
+
+    internal init(_ modelCollection: ModelCollection) {
+        self.represented = modelCollection
+    }
+
+    // MARK: NestedModelCollection
```

Comment:
Ah yeah good catch, I intended to have this wrapper to just be an implementation detail of `asNested()` which already does the catch. But really this class can be `private` rather than internal which should avoid it getting mistakenly used even within pilot.